### PR TITLE
[presto][DRAFT] Add codec interface

### PR DESCRIPTION
**Pull request by propene**
Add codec interface into presto-orc for supporting customized codec.

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/CompressionKind.java [hunk 1]
```diff
@@ -15,5 +15,5 @@
 
 public enum CompressionKind
 {
-    UNCOMPRESSED, ZLIB, SNAPPY
+    UNCOMPRESSED, ZLIB, SNAPPY, ENCRYPTED
```
Comment (martint) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/CompressionKind.java]: This conflicts with the specification of the ORC format: https://orc.apache.org/docs/file-tail.html

**Comment by martint** [comment]
The ORC format defines a well-known set of compression schemes. Given that adding a new requires changing the code, I'm not sure I understand how making it dynamically extensible would work. Specifically, you can't add a new codec without modifying the enum that contains the list of supported schemes.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/CompressionKind.java (hunk 1 above)
Comment (wdong87) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/CompressionKind.java]: @martint, Presto defined its own CompressionKind. Do we have to have exact match between this one and the Apache one? For example, 'UNCOMPRESSED' is only roughly mapped to 'NONE' there. Thanks!

**Comment by wdong87** [comment]
@martint, if Presto is flexible enough to take more enum options for CompressionKind than apache, it would be easier for us to do more innovations with some new formats. The dynamically extensible logic also helps us to plug in our custom codec even if there is a similar one in open source, for example, ZSTD will be a good example. Thoughts?

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/CompressionKind.java (hunk 1 above)
Comment (martint) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/CompressionKind.java]: > @martint, Presto defined its own CompressionKind. Do we have to have exact match between this one and the Apache one? For example, 'UNCOMPRESSED' is only roughly mapped to 'NONE' there. Thanks!

Ours is a subset. The names don't mean anything -- what matters is the number associated with each enum entry:

```
enum CompressionKind {
 NONE = 0;
 ZLIB = 1;
 SNAPPY = 2;
 LZO = 3;
 LZ4 = 4;
 ZSTD = 5;
}
```

Once we add support for LZO/LZ4/ZSTD we'll add the corresponding constants

**Comment by martint** [comment]
> @martint, if Presto is flexible enough to take more enum options for CompressionKind than apache, it would be easier for us to do more innovations with some new formats. The dynamically extensible logic also helps us to plug in our custom codec even if there is a similar one in open source, for example, ZSTD will be a good example. Thoughts?

It's not really about Presto being flexible. The enum values are part of the ORC format specification. Changing that would mean breaking compatibility (potentially, in the future, as the open specification continues to evolve)

**Comment by martint** [comment]
BTW, I'm still confused about how you expect this to work. The compression format is for the whole file. From the ORC spec:

> If the ORC file writer selects a generic compression codec (zlib or snappy), every part of the ORC file except for the Postscript is compressed with that codec.

So, it doesn't seem possible to encrypt only certain columns.

Also, since compression and encryption are orthogonal, how would you get ZLIB+ENCRYPTION vs LZ4+ENCRYPTION, etc.

**Comment by martint** [comment]
Closing, since it doesn't seem viable without major changes to the ORC format.

**Pull request closed without merging**